segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeperstorage/LogStorageManager.java

```diff
@@ -0,0 +1,701 @@
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.api.transaction.CuratorOp;
+import org.apache.curator.framework.api.transaction.CuratorTransactionResult;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.data.Stat;
+
+/**
+ * This class represents the manager for StorageLedgers.
+ * It is responsible for creating and deleting storage ledger objects in bookkeeper.
+ * This objects interacts with Zookeeper through Async curator framework to manage the metadata.
+ */
+@Slf4j
+class LogStorageManager {
+    private static final BookKeeper.DigestType LEDGER_DIGEST_TYPE = BookKeeper.DigestType.MAC;
+
+    private final BookKeeperStorageConfig config;
+    private final CuratorFramework zkClient;
+
+    @GuardedBy("this")
+    private final ConcurrentHashMap<String, LogStorage> ledgers;
+    private BookKeeper bookkeeper;
+    private long containerEpoch;
+
+    LogStorageManager(BookKeeperStorageConfig config, CuratorFramework zkClient) {
+        Preconditions.checkNotNull(config, "config");
+        Preconditions.checkNotNull(zkClient, "zkClient");
+
+        this.config = config;
+        this.zkClient = zkClient;
+        ledgers = new ConcurrentHashMap<>();
+    }
+
+    //region Storage API
+
+    /**
+     * Creates a LogStorage at location streamSegmentName.
+     *
+     * @param streamSegmentName Full name of the stream segment
+     * @return A new instance of LogStorage if one does not already exist
+     * @Param bookkeeper the BK instance
+     * @Param zkClient
+     */
+    public LedgerData create(String streamSegmentName) throws StreamSegmentException {
+        log.info("Creating segment {}", streamSegmentName);
+        LogStorage logStorage = new LogStorage(this, streamSegmentName, 0, this.containerEpoch, 0);
+
+        /* Create the node for the segment in the ZK. */
+        try {
+            zkClient.create()
+                    .forPath(getZkPath(streamSegmentName), logStorage.serialize());
+            synchronized (this) {
+                ledgers.put(streamSegmentName, logStorage);
+            }
+            LedgerData data = createLedgerAt(streamSegmentName, 0);
+            /** Add the ledger to the local cache.*/
+            synchronized (this) {
+                this.ledgers.get(streamSegmentName).addToList(0, data);
+            }
+            return data;
+        } catch (Exception exc) {
+            translateZKException(streamSegmentName, exc);
+            return null;
+        }
+    }
+
+
+    /**
+     * Fences an segment which already exists.
+     * Steps involved:
+     * 1. Read the latest ledger from ZK
+     * 2. Try to open it for write
+     * 3. If it is fenced, create a new one.
+     *
+     * @param streamSegmentName name of the segment to be fenced.
+     */
+    public LogStorage fence(String streamSegmentName) throws StreamSegmentException {
+        boolean tryAgain = true;
```

Comment:
This would be a bit cleaner if there were a private method that did the inner part and managed try again.

Reply:
Fixed. Refactored.